File: scripts/populate-municipalities.js
import { db } from '../server/db.js';
import { brazilianMunicipalities } from '../shared/schema.js';
import { sql } from 'drizzle-orm';

// Função para buscar e popular todos os municípios brasileiros
async function populateAllMunicipalities() {
  try {
    console.log('Buscando dados dos municípios da API do IBGE...');
    
    // Buscar todos os municípios do Brasil
    const response = await fetch('https://servicodados.ibge.gov.br/api/v1/localidades/municipios');
    const municipalities = await response.json();
    
    console.log(`Total de municípios encontrados: ${municipalities.length}`);
    
    // Limpar tabela antes de inserir novos dados
    await db.delete(brazilianMunicipalities);
    console.log('Tabela de municípios limpa.');
    
    // Processar em lotes para evitar problemas de performance
    const batchSize = 100;
    let processed = 0;
    
    for (let i = 0; i < municipalities.length; i += batchSize) {
      const batch = municipalities.slice(i, i + batchSize);
      
      const municipalityData = batch.map(municipality => ({
        name: municipality.nome,
        stateCode: municipality.microrregiao.mesorregiao.UF.sigla,
        ibgeCode: municipality.id.toString()
      }));
      
      await db.insert(brazilianMunicipalities).values(municipalityData);
      processed += batch.length;
      
      console.log(`Processados ${processed}/${municipalities.length} municípios...`);
    }
    
    console.log('✅ Todos os municípios foram inseridos com sucesso!');
    
    // Mostrar estatísticas por estado
    const stats = await db
      .select({
        stateCode: brazilianMunicipalities.stateCode,
        count: sql`count(*)`
      })
      .from(brazilianMunicipalities)
      .groupBy(brazilianMunicipalities.stateCode)
      .orderBy(brazilianMunicipalities.stateCode);
    
    console.log('\n📊 Municípios por estado:');
    stats.forEach(stat => {
      console.log(`${stat.stateCode}: ${stat.count} municípios`);
    });
    
  } catch (error) {
    console.error('❌ Erro ao popular municípios:', error);
  } finally {
    process.exit(0);
  }
}

// Executar o script
populateAllMunicipalities();

File: scripts/populate_municipalities.py
#!/usr/bin/env python3
import json
import os
import psycopg2
from psycopg2.extras import execute_values

def populate_municipalities():
    try:
        print("Carregando dados dos municípios...")
        
        # Carregar dados do arquivo JSON
        with open('/tmp/municipios.json', 'r', encoding='utf-8') as file:
            municipalities = json.load(file)
        
        print(f"Total de municípios encontrados: {len(municipalities)}")
        
        # Conectar ao banco de dados
        conn = psycopg2.connect(os.environ['DATABASE_URL'])
        cur = conn.cursor()
        
        # Limpar tabela antes de inserir novos dados
        cur.execute("DELETE FROM brazilian_municipalities")
        print("Tabela de municípios limpa.")
        
        # Preparar dados para inserção
        municipality_data = []
        for municipality in municipalities:
            municipality_data.append((
                municipality['nome'],
                municipality['microrregiao']['mesorregiao']['UF']['sigla'],
                str(municipality['id'])
            ))
        
        # Inserir em lotes para melhor performance
        batch_size = 1000
        total_inserted = 0
        
        for i in range(0, len(municipality_data), batch_size):
            batch = municipality_data[i:i + batch_size]
            execute_values(
                cur,
                "INSERT INTO brazilian_municipalities (name, state_code, ibge_code) VALUES %s",
                batch,
                template=None,
                page_size=100
            )
            total_inserted += len(batch)
            print(f"Inseridos {total_inserted}/{len(municipality_data)} municípios...")
        
        # Confirmar transação
        conn.commit()
        print("✅ Todos os municípios foram inseridos com sucesso!")
        
        # Mostrar estatísticas por estado
        cur.execute("""
            SELECT state_code, COUNT(*) as count 
            FROM brazilian_municipalities 
            GROUP BY state_code 
            ORDER BY state_code
        """)
        
        stats = cur.fetchall()
        print("\n📊 Municípios por estado:")
        for state_code, count in stats:
            print(f"{state_code}: {count} municípios")
        
        # Total final
        cur.execute("SELECT COUNT(*) FROM brazilian_municipalities")
        total = cur.fetchone()[0]
        print(f"\n🎉 Total de {total} municípios inseridos no banco de dados!")
        
    except Exception as error:
        print(f"❌ Erro ao popular municípios: {error}")
        if 'conn' in locals():
            conn.rollback()
    finally:
        if 'cur' in locals():
            cur.close()
        if 'conn' in locals():
            conn.close()

if __name__ == "__main__":
    populate_municipalities()